pkg/tokenreviews/apiserver.go

```diff
@@ -0,0 +1,187 @@
+package apiserver
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	kauthv1 "k8s.io/api/authentication/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apiserver/pkg/registry/rest"
+	genericapiserver "k8s.io/apiserver/pkg/server"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/kubernetes/pkg/registry/authentication/tokenreview"
+
+	oauthclients "github.com/openshift/client-go/oauth/clientset/versioned"
+	oauthinformer "github.com/openshift/client-go/oauth/informers/externalversions"
+	userclients "github.com/openshift/client-go/user/clientset/versioned"
+	userinformer "github.com/openshift/client-go/user/informers/externalversions"
+
+	bootstrap "github.com/openshift/library-go/pkg/authentication/bootstrapauthenticator"
+	"github.com/openshift/oauth-apiserver/pkg/serverscheme"
+	tokenvalidation "github.com/openshift/oauth-apiserver/pkg/tokenreviews/tokenvalidation"
+	"github.com/openshift/oauth-apiserver/pkg/tokenreviews/tokenvalidation/usercache"
+	tokenvalidators "github.com/openshift/oauth-apiserver/pkg/tokenreviews/tokenvalidation/validators"
+)
+
+const (
+	defaultInformerResyncPeriod     = 10 * time.Minute
+	minimumInactivityTimeoutSeconds = 300
+)
+
+type ExtraConfig struct {
+	ServiceAccountMethod string
+
+	makeV1Storage sync.Once
+	v1Storage     map[string]rest.Storage
+	v1StorageErr  error
+
+	postStartHooks map[string]genericapiserver.PostStartHookFunc
+
+	AccessTokenInactivityTimeout time.Duration
+}
+type TokenReviewsAPIServerConfig struct {
+	GenericConfig *genericapiserver.RecommendedConfig
+	ExtraConfig   ExtraConfig
+}
+
+type TokenReviewsAPIServer struct {
+	GenericAPIServer *genericapiserver.GenericAPIServer
+}
+
+type completedConfig struct {
+	GenericConfig             genericapiserver.CompletedConfig
+	ExtraConfig               *ExtraConfig
+	kubeAPIServerClientConfig *restclient.Config
+}
+
+type CompletedConfig struct {
+	// Embed a private pointer that cannot be instantiated outside of this package.
+	*completedConfig
+}
+
+// Complete fills in any fields not set that are required to have valid data. It's mutating the receiver.
+func (c *TokenReviewsAPIServerConfig) Complete() completedConfig {
+	cfg := completedConfig{
+		GenericConfig:             c.GenericConfig.Complete(),
+		ExtraConfig:               &c.ExtraConfig,
+		kubeAPIServerClientConfig: c.GenericConfig.ClientConfig,
+	}
+
+	return cfg
+}
+
+// New returns a new instance of TokenReviewsAPIServer from the given config.
+func (c completedConfig) New(delegationTarget genericapiserver.DelegationTarget) (*TokenReviewsAPIServer, error) {
+	genericServer, err := c.GenericConfig.New("tokenreviews.openshift.io-apiserver", delegationTarget)
+	if err != nil {
+		return nil, err
+	}
+
+	s := &TokenReviewsAPIServer{
+		GenericAPIServer: genericServer,
+	}
+
+	v1Storage, postStartHooks, err := c.V1RESTStorage()
```

Comment:
We want this part of the storage added at `apis/oauth.openshift.io/v1/tokenaccessreviews`, but accepting the kube-apiserver format.  

See https://github.com/openshift/generic-admission-server/blob/master/pkg/registry/admissionreview/admission_review.go#L33-L45 for how to accept one version of a think in REST handler and serve from a different spot.  Very close now.